operator/v1/types_ingress.go

```diff
@@ -225,6 +233,97 @@ var (
 	DNSReadyIngressConditionType = "DNSReady"
 )
 
+// SecuritySpec defines the settings for securing IngressController connections.
+type SecuritySpec struct {
+	// Profile is one of three possible security configurations from [1]:
+	//
+	// [1] https://wiki.mozilla.org/Security/Server_Side_TLS#Recommended_configurations
+	//
+	// If unset, the default is "intermediate".
+	//
+	// +optional
+	Profile *SecurityProfileType `json:"profile,omitempty"`
+	// CustomSettings defines the schema for customizing the security settings of
+	// IngressController connections.
+	//
+	// +optional
+	CustomSettings *CustomProfileSettings `json:"customSettings,omitempty"`
+}
+
+// SecurityProfileType defines a security profile.
+type SecurityProfileType string
+
+const (
+	// SecurityProfileOldType maps to:
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Old_backward_compatibility
+	SecurityProfileOldType SecurityProfileType = "old"
+	// SecurityProfileIntermediateType maps to:
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28default.29
+	SecurityProfileIntermediateType SecurityProfileType = "intermediate"
+	// SecurityProfileModernType maps to:
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility
+	SecurityProfileModernType SecurityProfileType = "modern"
+)
+
+// CustomProfileSettings defines the schema for a custom security profile.
+type CustomProfileSettings struct {
+	// Cyphers is used to specify the cipher algorithms that are negotiated
+	// during the SSL/TLS handshake with an IngressController. Each cipher must
+	// be an explicit, colon-delimited list of ciphers.
+	//
+	// If unset, the intermediate ciphersuite [1] is used:
+	//
+	// [1] https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28default.29
+	//
+	// +optional
+	Cyphers *string `json:"cyphers,omitempty"`
+	// SecurityProtocol is used to specify one or more encryption protocols
+	// that are negotiated during the SSL/TLS handshake with the IngressController.
+	//
+	// If unset, TLSv1.0-TLSv1.3 are used.
+	//
+	// +optional
+	SecurityProtocol *SecurityProtocol `json:"securityProtocol,omitempty"`
+}
+
+// SecurityProtocol defines one or more security protocols used by
+// an IngressController to secure network connections.
+type SecurityProtocol struct {
+	// MinimumVersion enforces use of SecurityProtocolVersion or newer on
+	// SSL connections instantiated from an IngressController.
+	//
+	// If unset, TLSv1.0-TLSv1.3 are used.
```

Comment:
Should that be, "If unset, TLSv1.0 is used."?

Is the defaulting logic the same if one field is specified and the other is not versus leaving both unspecified?

Reply:
Nope. If unset, all protocols are enabled instead of SSLv3. [Code example](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L61).

Reply:
(Note: This is all moot if we decide not to use a range, per one of my other comments.)

This is the godoc for `minimumVersion`, not for the combination of `minimumVersion` and `maximumVersion`.  The two fields together can default to `TLSv1.0-TLSv1.3`, but each individual field must default to some single value of type `SecurityProtocolVersion`.

I'm guessing that the intention is that `minimumVersion` default to `TLSv1.0` and `maximumVersion` to `TLSv1.3`, but the defaulting of one field could be dependent on the value of the other.  For example, if one field were specified and the other were not, the default for the unspecified field _could_ be defined as the value of the specified field—in this case, specifying one but not the other would mean "exactly the specified version (not a range)".

Indeed, I can think of at least one edge case where we cannot default the values independently: if `maximumVersion` is `SSLv3`, then `minimumVersion` cannot default to `TLSv1.0`.

Alternatively, we could sidestep the issue by defining a range type with required minimum and maximum values, and using an optional field of that range type.

Reply:
@Miciah I included additional godocs to describe the behavior of `MinimumVersion` and `MaximumVersion`. PTAL and let me know your thoughts.